src/NodeApi/Engines/NodejsEnvironment.cs

```diff
@@ -0,0 +1,205 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.JavaScript.NodeApi.Interop;
+using static Microsoft.JavaScript.NodeApi.JSNativeApi.Interop;
+
+namespace Microsoft.JavaScript.NodeApi.Engines;
+
+/// <summary>
+/// A Node.js runtime environment with a dedicated main execution thread.
+/// </summary>
+/// <remarks>
+/// Multiple Node.js environments may be created (concurrently) in the same process. Each
+/// environment instance has its own dedicated execution thread. Except where otherwise documented,
+/// all interaction with the environment and JavaScript values associated with the environment MUST
+/// be executed on the environment's thread. Use the
+/// <see cref="NodejsEnvironment.SynchronizationContext" /> to switch to the thread.
+/// </remarks>
+public sealed class NodejsEnvironment : IDisposable
+{
+    private readonly JSValueScope _scope;
+    private readonly Thread _thread;
+    private readonly TaskCompletionSource<bool> _completion = new();
+
+    public static explicit operator napi_env(NodejsEnvironment environment) =>
+        (napi_env)environment._scope;
+    public static implicit operator JSValueScope(NodejsEnvironment environment) =>
+        environment._scope;
+
+    internal NodejsEnvironment(NodejsPlatform platform, string? mainScript)
+    {
+        JSValueScope scope = null!;
```

Comment:
>JSValueScope scope = null!; [](http://example.com/codeflow?start=8&length=27)

Ideally we must not use the value scope like this. It must be specific to the JS thread.

Reply:
Overall, we better to keep exactly the same initialization sequence regardless of the runtime we use.

Reply:
The `JSValueScope` is constructed from within the JS thread. It is only captured and assigned to `_scope` outside the lambda to satisfy the compiler because `_scope` is `_readonly` and therefore must be assigned from the constructor thread.

I don't understand what you mean by "keep the same initialization sequence regardless of the runtime". Other runtimes have different requirements for initialization and threading.